File: 54.螺旋矩阵.cpp
/*
 * @lc app=leetcode.cn id=54 lang=cpp
 * @lcpr version=30203
 *
 * [54] 螺旋矩阵
 */

// @lc code=start
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        vector<int> res;
        if (matrix.empty()) {
            return res;
        }
        int m = matrix.size(), n = matrix[0].size();
        int left_bound = 0, right_bound = n - 1, up_bound = 0, down_bound = m - 1;
        while (left_bound <= right_bound && up_bound <= down_bound) {
            for (int i = left_bound; i <= right_bound && up_bound <= down_bound; i++) {
                res.push_back(matrix[up_bound][i]);
            }
            up_bound++;
            for (int j = up_bound; j <= down_bound && left_bound <= right_bound; j++) {
                res.push_back(matrix[j][right_bound]);
            }
            right_bound--;
            for (int p = right_bound; p >= left_bound && up_bound <= down_bound; p--) {
                res.push_back(matrix[down_bound][p]);
            }
            down_bound--;
            for (int q = down_bound; q >= up_bound && left_bound <= right_bound; q--) {
                res.push_back(matrix[q][left_bound]);
            }
            left_bound++;
        }
        return res;
    }
};
// @lc code=end



/*
// @lcpr case=start
// [[1,2,3],[4,5,6],[7,8,9]]\n
// @lcpr case=end

// @lcpr case=start
// [[1,2,3,4],[5,6,7,8],[9,10,11,12]]\n
// @lcpr case=end

 */

File: 54.螺旋矩阵.java
/*
 * @lc app=leetcode.cn id=54 lang=java
 * @lcpr version=30305
 *
 * [54] 螺旋矩阵
 */

// @lc code=start

import java.util.ArrayList;
import java.util.List;

class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> res = new ArrayList<>();
        int m = matrix.length, n = matrix[0].length;
        int up = 0, down = m - 1, left = 0, right = n - 1;
        while (up <= down || left <= right) {
            // 在第 up 行遍历，关注左右
            for (int x = left; x <= right && up <= down; x++) {
                res.add(matrix[up][x]);
            }
            up++;

            // 在第 right 列进行遍历,关注上下
            for (int y = up; y <= down && left <= right; y++) {
                res.add(matrix[y][right]);
            }
            right--;

            // 在第 up 行遍历，关注左右
            for (int x = right; x >= left && up <= down; x--) {
                res.add(matrix[down][x]);
            }
            down--;

            // 在第 right 列进行遍历,关注上下
            for (int y = down; y >= up && left <= right; y--) {
                res.add(matrix[y][left]);
            }
            left++;
        }
        return res;
    }
}
// @lc code=end



/*
// @lcpr case=start
// [[1,2,3],[4,5,6],[7,8,9]]\n
// @lcpr case=end

// @lcpr case=start
// [[1,2,3,4],[5,6,7,8],[9,10,11,12]]\n
// @lcpr case=end

 */

